channeldb/payments.go

```diff
@@ -0,0 +1,211 @@
+package channeldb
+
+import (
+	"github.com/roasbeef/btcutil"
+	"io"
+	"github.com/roasbeef/btcd/wire"
+	"github.com/boltdb/bolt"
+	"encoding/binary"
+	"bytes"
+)
+
+var (
+	// invoiceBucket is the name of the bucket within the database that
+	// stores all data related to payments.
+	// Within the payments bucket, each invoice is keyed by its invoice ID
+	// which is a monotonically increasing uint64.
+	// BoltDB sequence feature is used for generating monotonically increasing
+	// id.
+	paymentBucket = []byte("payments")
+)
+
+type OutgoingPayment struct {
+	Invoice
+	// Total fee paid
+	Fee btcutil.Amount
+	// Path including starting and ending nodes
+	Path [][]byte
+	TimeLockLength uint64
+	// We probably need both RHash and Preimage
+	// because we start knowing only RHash
+	RHash [32]byte
+}
+
+func validatePayment(p *OutgoingPayment) error {
+	err := validateInvoice(&p.Invoice)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// AddPayment adds payment to DB.
+// There is no checking that payment with the same hash already exist.
+func (db *DB) AddPayment(p *OutgoingPayment) error {
+	err := validatePayment(p)
+	if err != nil {
+		return err
+	}
+	// We serialize before writing to database
+	// so no db access in the case of serialization errors
+	b := new(bytes.Buffer)
+	err = serializeOutgoingPayment(b, p)
+	if err != nil {
+		return err
+	}
+	paymentBytes := b.Bytes()
+	return db.Update(func (tx *bolt.Tx) error {
+		payments, err := tx.CreateBucketIfNotExists(paymentBucket)
+		if err != nil {
+			return err
+		}
+		paymentId, err := payments.NextSequence()
+		if err != nil {
+			return err
+		}
+		// We use BigEndian for keys because
+		// it orders keys in ascending order
+		paymentIdBytes := make([]byte, 8)
+		binary.BigEndian.PutUint64(paymentIdBytes, paymentId)
+		err = payments.Put(paymentIdBytes, paymentBytes)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+}
+
+// FetchAllPayments returns all outgoing payments in DB.
+func (db *DB) FetchAllPayments() ([]*OutgoingPayment, error) {
+	var payments []*OutgoingPayment
+	err := db.View(func (tx *bolt.Tx) error {
+		bucket := tx.Bucket(paymentBucket)
+		if bucket == nil {
+			return nil
```

Comment:
I think `ErrNoPaymentsCreated ` should be returned here instead of a `nil` error. I see you create the type but haven't yet used it anywhere within this PR, so maybe this is just due to the PR not being _fully_ complete yet.